pkg/operator/controller/nodeobservabilityrun/controller.go

```diff
@@ -18,58 +18,124 @@ package nodeobservabilityruncontroller
 
 // Reconcile manages NodeObservabilityRuns
-func (r *NodeObservabilityRunReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (r *NodeObservabilityRunReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
 	instance := &nodeobservabilityv1alpha1.NodeObservabilityRun{}
-	err := r.Get(ctx, req.NamespacedName, instance)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			return ctrl.Result{}, nil
-		}
-		return ctrl.Result{}, err
+	err = r.Get(ctx, req.NamespacedName, instance)
+	if client.IgnoreNotFound(err) != nil {
+		r.Log.Error(err, "failed to get NodeObservabilityRun")
+		return
 	}
 
 	if finished(instance) {
-		return ctrl.Result{}, nil
+		return
 	}
 
+	defer func() {
+		errUpdate := r.Status().Update(ctx, instance, nil)
+		if errUpdate != nil {
+			err = utilerrors.NewAggregate([]error{err, errUpdate})
+		}
+	}()
+
 	if inProgress(instance) {
-		err = r.handleInProgress(instance)
+		var requeue bool
+		requeue, err = r.handleInProgress(instance)
+		if requeue {
+			return ctrl.Result{RequeueAfter: pollingPeriod}, err
+		}
+		t := metav1.Now()
+		instance.Status.FinishedTimestamp = &t
+		return
+	}
+
+	targets, err := r.startRun(ctx, instance)
+	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	err = r.startRun(instance)
+	instance.Status.Agents = targets
+	t := metav1.Now()
+	instance.Status.StartTimestamp = &t
 
-	return ctrl.Result{}, err
+	// one cycle takes cca 30s
+	return ctrl.Result{RequeueAfter: time.Second * 30}, err
 }
 
-func (r *NodeObservabilityRunReconciler) handleInProgress(instance *nodeobservabilityv1alpha1.NodeObservabilityRun) error {
-	return nil
+func (r *NodeObservabilityRunReconciler) handleInProgress(instance *nodeobservabilityv1alpha1.NodeObservabilityRun) (requeue bool, err error) {
+	var code int
+	for _, agent := range instance.Status.Agents {
+		_, code, err = httpGetCall(fmt.Sprintf("http://%s:%d/status", agent.IP, agent.Port))
+		if err != nil {
+			// TODO: node consistently failing, restart whole process
+			return
+		}
+		if code == http.StatusConflict {
+			requeue = true
+			return
+		}
+	}
+
+	return
 }
 
-func (r *NodeObservabilityRunReconciler) startRun(instance *nodeobservabilityv1alpha1.NodeObservabilityRun) error {
-	return nil
+func (r *NodeObservabilityRunReconciler) startRun(ctx context.Context, instance *nodeobservabilityv1alpha1.NodeObservabilityRun) ([]nodeobservabilityv1alpha1.AgentNode, error) {
+	r.Log.V(3).Info("startRun", "instance", instance.Name)
+	endps, err := r.getAgentEndpoints(ctx)
+	if err != nil {
+		return nil, err
+	}
+	subset := endps.Subsets[0]
+	port := subset.Ports[0].Port
+
+	targets := []nodeobservabilityv1alpha1.AgentNode{}
+	for _, a := range subset.Addresses {
+		r.Log.V(3).Info("startRun - address for loop", "IP", a.IP)
+		body, code, err := httpGetCall(fmt.Sprintf("http://%s:%d/pprof", a.IP, port))
+		if err != nil || code != 200 {
```

Comment:
Open question: is there any gain in noting this IP down in a UnavailableTargets in the status, and eventually retrying it later within the handleInProgress?